[WIP] No-overlay mode enhancement

pliurh opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/ovn-kubernetes/ovn-kubernetes/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

All changes must adhere to this template to make it easy for reviewers
and preserve rationale/history behind every change
-->

## 📑 Description
<!-- Add a brief description of the pr -->

<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

## Additional Information for reviewers
<!--
What exactly did you change - you may also defer to information
contained in commit messages. At a bare minimum it's worth highlighting
which areas of the code were changed as it's easier to assign reviewers
-->

## ✅ Checks
<!-- Make sure your pr passes the CI checks and do check the following fields as needed - -->
- [ ] My code requires changes to the documentation
- [ ] if so, I have updated the documentation as required
- [ ] My code requires tests
- [ ] if so, I have added and/or updated the tests as required
- [ ] All the tests have passed in the CI <!-- If not leave a comment as to why the CI is red and if you need help understanding what's wrong -->

## How to verify it
<!--
Did you include unit tests? or end-to-end tests?
How can I manually verify that this patch achieves its objective
-->

zhaozhanqi commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,595 @@
+
+* Support no-overlay mode for the default network
+* Support no-overlay mode for layer-3 Cluster User Defined Networks (CUDNs) in
+  the VRF-lite mode .
```
Here montioned VRF-lite mode,  How about default VRF mode for CUDN?

zhaozhanqi replied:
OK,  please ignore this,  found the details explanation in the following part.

zhaozhanqi commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,595 @@
+
+* When OVN-Kubernetes supports BGP for UDNs, extend the UDN API to enable
+  no-overlay mode on the selected network
+* Support toggling no-overlay mode on/off for an existing network.
```
Support togging no-overlay mode on/off runtime by patch the CUDN resource, right?

pliurh replied:
The default network and UDNs. We can remove the API restriction if this use case makes sense.

zhaozhanqi commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,595 @@
+#### Example of the default network with no-overlay mode enabled
+
+For the cluster default network, the cluster admin shall create a
+ClusterUserDefinedNetwork CR with the reserved name `default` in the namespace where
```
is that possible enable no-overlay node for default network by CNO like `no-overlay-default-network: enable` and let CNO to create a default ClusterUserDefinedNetwork CR ?

zhaozhanqi commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,595 @@
+
+No-overlay mode for the default network must be enabled before OVN-Kubernetes
+pods start. Therefore, the following configuration manifests shall be created
+before deploying the OVN-Kubernetes pods.
```
so is that mean we need to re-created the ovn-kubernetes pods after creating the CR on existing cluster to enable the default network no-overlay?

pliurh replied:
No, the CUDN CR shall be created before ovn-k pods start.

ricky-rav replied:
I didn't get this either. So can the default network be configured to use no-overlay mode only at installation time? Or are we telling users to kill ovn-k pods after creating the CUDN CR for the default network? I'm not sure if either case is ideal :)

zhaozhanqi commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,595 @@
+transit switch anymore and all ease-west traffic will be naturally routed in the
+same path as north-south traffic through the provider network.
+
+In Shared Gateway Mode, at the OVN ClusterRouter, the pod to remote pod egress
```
OK, So the EIP pods egress outside, still overlay to egress node

pliurh replied:
I may have underestimated the complexity of supporting EIP in no-overlay mode. Using the overlay network to forward traffic from a pod to the egress node may not be the best approach.

ricky-rav commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,595 @@
+without being encapsulated in a Geneve tunnel and without being SNATed to the
+node IP.
+
+The only exception is for pod to other host traffic, to ensure nodePort services
```
Uhm here I don't follow. Are we keeping the SNAT so that we can allow traffic to nodeports on different nodes, or are we keeping the SNAT so that users won't expose pod IPs to the outside?

pliurh replied:
We keep the SNAT only for pod-to-remote_host traffic.

ricky-rav commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,595 @@
+      name: default
+    spec:
+      # use an empty selector to select all namespaces
+      namespaceSelector: {}
```
Should we invalidate all other values for `namespaceSelector`?

pliurh replied:
Maybe we can add a validation rule for this.

ricky-rav commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,595 @@
+Here's a configuration example:
+
+1. A user want to enable no-overlay mode for the blue network. The user has to
+   do all host modifications necessary via NMState to enslave an VLAN interface
```
```suggestion
   do all host modifications necessary via NMState to enslave a VLAN interface
```

ricky-rav commented on docs/okeps/okep-5259-no-overlay.md:
````diff
@@ -0,0 +1,595 @@
+                mode: filtered
+    ```
+
+1. The user have to advertise this pod network of the CUDN by BGP. The targetVRF
````
```suggestion
1. The user has to advertise this pod network of the CUDN by BGP. The targetVRF
```

ricky-rav commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,595 @@
+
+#### BGP Topology
+
+The deployment mode will be decided by the FRRConfiguration CR that is selected
```
To stress that it's up to the cluster admin:
```suggestion
The deployment mode will be decided by the FRRConfiguration CR that the user creates
```

BTW should we replace `user` with `cluster admin` all throughout the enhancement?

zhaozhanqi commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,595 @@
+
+#### IPsec
+
+Not Support.
```
so on ipsec cluster, there will `not supported` message shown when applying the no overlay CR?

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,595 @@
+The cross node east-west traffic in this user defined network will be isolated
+the by the VRFs in each node and the VLANs of the provider network. VRF-Lite is
+only available when using Local Gateway Mode, thus no-overlay mode user defined
+networks is also only available when using Local Gateway Mode.
```
I want to ensure the network isolation is implemented end-to-end (not only within a node but also in the provider network), similar to what we used to have in overlay mode. The node overlay mode is for east-west traffic; mixing routes for different UDNs in a single VRF could introduce a parental security vulnerability.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,597 @@
+      subnets:
+      - cidr: 10.10.0.0/16
+        hostSubnet: 24
+      encapsulation: None
```
UDN support is a future goal. We can validate it in the cluster manager and set a `status.condition` in the UDN CR to indicate that this configuration is not supported yet.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,597 @@
+ClusterUserDefinedNetwork CR with the reserved name `default` and set
+encapsulation to `None` to enable no-overlay mode before deploying
+OVN-Kubernetes. For this CR, OVN-K shall not create the corresponding
+NetworkAttachementDefinition CR. The subnet field must aligned with the value in
```
Sure, but the NAD doesn't have to be derived from the `default` CUDN unless that's our intention.

jcaamano replied:
I don't know what role the NAD (any NAD) has on this enhancement. I think it doesn't have any role or at least it is not mentioned.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,597 @@
+
+For the cluster default network, the cluster admin shall create a
+ClusterUserDefinedNetwork CR with the reserved name `default` and set
+encapsulation to `None` to enable no-overlay mode before deploying
```
I've also considered this option. I considered 2 points: 1) implementing a unified API for all ovn-k networks; and 2) planning for future support of migration between different encapsulation methods.

jcaamano replied:
Additionally, if we are going to use a CUDN for this, then we should describe what happens when:
- the `default` CUDN is created after OVN-Kubernetes has already started
- the `default` CUDN is deleted after OVN-Kubernetes has already started
- the `default` CUDN is created after OVN-Kubernetes has already started but while being down

So, agreeing with @trozet here, using a CUDN when it is not supposed to be a day-2 thing seems like an extra complication.

pliurh replied:
We need to decide whether to support changing the encapsulation for the default network during day-2 operations. If the answer is no, then I agree we shall use a flag instead of the reserved CR.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,595 @@
+
+The `spec` field of a ClusterUserDefinedNetwork CR is immutable. Therefore, the
+encapsulation configuration cannot be changed once a ClusterUserDefinedNetwork
+CR is created.
```
This is the current implementation. https://github.com/ovn-kubernetes/ovn-kubernetes/blob/682ffaf5d4550fd5dac1624ef53de7ee5d82416f/dist/templates/k8s.ovn.org_clusteruserdefinednetworks.yaml.j2#L537

ricky-rav replied:
Yeah, that's true. If this aligns with the expectations for this feature, all good :)

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,595 @@
+  egress. In the route advertisement feature, SNAT is disabled for all pod
+  egress traffic that leaves the node, regardless of whether the destination
+  route is BGP learnt or not. For users who only want to have no-overlay mode
+  but not want to expose podIPs to external, this behavior is not ideal. Instead
```
Why not. No-overlay is for east-west, exposing pod IP to external is for north-south.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,597 @@
+new field will enable control over the network encapsualtion behavior with the
+following options:
+
+1. **Encapsulation Configuration**:
```
Using `VXLAN` for EVPN could be misleading. Since OVN itself also supports using vxlan as overlay protocol. Maybe, we could use `EVPN_VXLAN` or `EVPN_MPLS` instead.

jcaamano replied:
If the concern is overwriting a user provided value (valid concern in my opinion), then we can require the user to set that field to the required value instead, otherwise we don't reconcile the EVPN CRD. 

In terms of what value should that be: perhaps `None` if the network is going to be configured in the same way as no overlay mode without EVPN, or perhaps we can require the user to set special value `EVPN_VXLAN` even if the network is not configured differently as a form of mutual acknowledgement or coordination, example: with a special value maybe we know we need to wait for something else to happen to reconfigure the network.

trozet replied:
@pliurh I think we still need to figure this one out as well

trozet replied:
I think the problem with that is if someone selects multiple EVPNs...hmm, maybe this is a flaw in my EVPN CRD as well. Someone cannot select more than 1 CUDN for MAC-VRF (L2 domain) that wouldn't make any sense connecting multiple L2 domains to the same L2 domain. 

They could select more than 1 CUDN for an IP-VRF though, since that is pure routing, as long as they don't have overlapping subnets.

Maybe your suggestion is a very good one. Then I can have several CUDNs select the same EVPN CRD, and if they are layer 3, they will just use the ip-vrf ID specified there. If the CUDN specifies a layer 2 and evpnSelector, and that EVPN CR already has another layer2 network configured for it, then it reports error status to the CUDN. That will tell the user that CUDN network failed, rather than sending that status in the EVPN CR.

I like this approach! @jcaamano wdyt?

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,597 @@
+all other nodes across the cluster. In this mode, users can enable the
+no-overlay mode without relying on external BGP routes.
+
+##### External Route Reflectors for Larger Clusters
```
Do you mean using eBGP within the cluster? Is this a valid use case? Also, BGP implementation doesn't support such a configuration. Routes to a podNetwork can only be advertised together.
https://github.com/ovn-kubernetes/ovn-kubernetes/blob/682ffaf5d4550fd5dac1624ef53de7ee5d82416f/dist/templates/k8s.ovn.org_routeadvertisements.yaml.j2#L546

trozet replied:
You don't need to have iBGP peering relationships between the nodes. You can use eBGP with multihop +loopback. You put each node in its own AS and then you can either peer full mesh style, or segment peering how you want to. This is useful in a cluster with a large amount of worker nodes, where you cant afford to do full mesh.

pliurh replied:
I see. However, I am uncertain whether our current BGP implementation supports such a configuration. OVN-K shall generate the frrconfiguration CR for each node, using their respective original frrconfiguration CRs. @jcaamano WDYT?

jcaamano replied:
We already can generate frrconfiguration CR for each node from a frrconfiguration CR for each node. So I don't see anything that prevents this configuration right now.
But this, as many others, would be a configuration combination that we are not doing integration testing for. Given the amount of alternatives, we need to either restrict the set that we support in some way, and add to them over time, or trust they just work because we don't see a reason they wouldn't. 
I see the extra possibilities of eBGP but I don't see how it would be preferred over a route reflector.
I don't have anything against mentioning it as an alternative. But if we think that we need to act on it in any other way, then we need to be more explicit.

trozet replied:
It would be preferred over a route reflector where:
1. There is no external bgp speaker.  Then something needs to designate at least 2 RRs (for HA) and now you need  BGP peering sessions from all nodes to both. Managing the RRs might become cumbersome. With eBGP you can break up large chunks of nodes into AS and then use multihop and build out the topology as you see fit. There is more flexibility.
2. There is an external bgp speaker, but they do not want to use iBGP. A lot of users will not want to use iBGP. I think it will be more common to use eBGP. It provides clear AS boundaries and allows for some extra functionality (AS path prepend, MED, filtering on as path). Also the next hop with iBGP will not set to be the worker node IP, unless it is peered "with next-hop-self".

pliurh replied:
Yeah, this is a valid use case. I added an eBGP topology example to the document. PTAL.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,597 @@
+
+##### Full-Mesh iBGP across the Cluster
+
+The FRR instance on each node maintains a full mesh BGP peer relationship with
```
This can be done by adding all the node router IDs into the same FrrConfiguration. We don't need to make any changes to the BGP implementation. It already works today.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,597 @@
+#### Create A ClusterUserDefinedNetwork in No-overlay Mode
+
+For user defined networks, to ensure the traffic isolation, no-overlay mode can
+only be enabled when the route advertisement is configured in a VFR-Lite mode.
```
Technically, it can be done. But it will break the end-to-end isolation. It could introduce a potential security vulnerability. No-overlay mode should maintain the same level of isolation as overlay mode. But it's just my thought. I'm okay with it if we want the user to decide.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,595 @@
+  but not want to expose podIPs to external, this behavior is not ideal. Instead
+  of disabling SNAT for all the pod-to-external traffic, OVN-K shall adjust its
+  SNAT behavior so that pod IPs are exposed for traffic routed via BGP, but SNAT
+  is applied for non-BGP-routed egress traffic.
```
Actually, we've decided to make this change to the BGP implementation. It's for the BGP feature, not only for no-overlay mode. Once that change is done, we can remove these sentences from this doc.

jcaamano commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,598 @@
+    spec:
+      # nodeSelector must be empty, since we don't support a network in a overlay and no-overlay hybrid mode.
+      nodeSelector: {}
+      frrConfigurationSelector:
+        matchLabels:
+          network: default
```
Missing details on the reasoning on what FRRConfiguration should be selected. Why match labels `network: default`?

pliurh replied:
The no-overlay mode relies on BGP routers to exchange the routes to pod subnets. We use the selected FRRConfiguration to connect to the remote BGP router. The label is just an example.

jcaamano commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,598 @@
+
+![No-overlay topology](../images/no_overlay_topology.jpg)
+
+#### No SNAT in no-overlay mode for egress traffic
```
It is previously mentioned that SNAT will be kept in place for non-BGP next hops, but I am missing implementation details about this.

jcaamano commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,598 @@
+
+1. OVN-kubernetes cluster manager watches the `default` CUDN CR. If
+   `spec.layer3.encapsulation` equals `None`, cluster manager will create a
+   `RouteAdvertisement` CR which will advertise pod subnets in the default
```
So this RouteAdvertisements would be owned by ovnk. If this is the case then I think it is worthwhile to have some implementation details on how this is reconciled and how this status is reflected back to the CUDN crd status. I ask myself questions like: what happens if someone deletes this RouteAdvertisements?

jcaamano replied:
On the other side, why does it it need to be created/owned by OVNK? Why isn't the user responsible for it?
Or, why don't we orchestrate no overlay mode for the default network from CNO? It might make sense to create the necessary FRRConfigurations, RouteAdvertisments and pass appropriate arguments to OVNK from there.

trozet replied:
yeah I'm very wary of automatically creating this CR. I'd prefer for default network we just use a config flag. Unless you can tell us more about the plans for dynamically turning it on and off for default network in the future. I would argue it probably makes sense to force the user to re-roll out the ovnk daemonset, since this is like a one time config change that isnt going to change often.

jcaamano replied:
To handle MTU issues, we have the MTU migration process. That would be enough for the default network. At some point we probably would need to find a way to loop in the UDNs in that MTU migration process, regardless of this enhancement.

jcaamano commented on docs/okeps/okep-5259-no-overlay.md:
````diff
@@ -0,0 +1,597 @@
+    ```
+
+1. OVN-kubernetes cluster manager watches the `default` UDN CR. If
+   `spec.layer3.encapsulation` equals `None`, cluster manager will create a
````
Just noting that this could be done on a different RA and configuration would be aggregated.

jcaamano commented on docs/okeps/okep-5259-no-overlay.md:
````diff
@@ -0,0 +1,598 @@
+                mode: filtered
+    ```
+
+1. The cluster admin has to advertise this pod network of the CUDN by BGP. The
+   targetVRF is set to auto, meaning routes to the podNetworks will only be
+   advertised within the corresponding VRF.
````
So if I am understanding this right, OVNK will create a RouteAdvertisements for the default network, but not for CUDNs?

pliurh replied:
I've updated the steps. Users shall create the RA.

jcaamano commented on docs/okeps/okep-5259-no-overlay.md:
````diff
@@ -0,0 +1,598 @@
+      encapsulation: None
+```
+
+### Implementation Details
````
I am missing implementation details describing what OVNK components react to the new API and how.

pliurh replied:
I will add.

jcaamano reviewed (commented):
Overall 
- I am still not convinced with the default network workflow
- Missing some implementation details

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,598 @@
+integrate seamlessly with existing BGP-capable network, achieve maximum network
+performance and simplify troubleshooting.
+
+### Story 3: enable no-overlay mode without an external BGP router
```
This is implemented by using a specific FrrConfiguration, which configures all nodes as BGP peers. It's already available with the current BGP implementation. I can give a configuration example for this setup.

zhaozhanqi commented:
so like the no-overlay mode for default network also rely on the BGP ( frr-k8s) , in case frr-k8s is stop working or not stable. The traffic between different nodes will also be affected. but I hope my my concern is unnecessary :)

jcaamano commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,598 @@
+* This enhancement does not aim to change the existing CUDN lifecycle
+  management. The user must ensure that the CUDN CRs are correctly managed
+  according to the existing lifecycle management practices.
+* This enhancement does not aim to implement the no-overlay mode with the
+  centralized OVN architecture.
+* This enhancement does not aim to implement the no-overlay mode for layer-2 or
+  localnet type of networks.
+* This enhancement does not aim to support an overlay and no-overlay hybrid
+  network.
```
BTW, I can really think a few reasons why Layer-2 support is more challenging. 

But what I really tried to say is that its better to write them down in the enhancement so we can comment on those if needed.

jcaamano commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,597 @@
+integrate seamlessly with existing BGP-capable network, achieve maximum network
+performance and simplify troubleshooting.
+
+### Story 3: enable no-overlay mode without an external BGP router
```
Got it now.  I consider it default behavior (although it isn't), let's call it loose isolation mode.

Right, it is a problem and I don't have a straight answer right now.

First we would need to make an assumption of what is the next hop for that traffic (if we don't want it to be directly the cluster node). Sans an extra API, we would have to assume it is the default gateway.

For SGW mode, we have the option of not importing other networks routes into the GR.

I don't think we have an equivalent alternative in LGW mode. I thought about FRR facilities for blacklisting/rewriting imported/received routes, because I think that is the best place to solve this, but couldn't find/come up with anything so far. 

So the only thing I can think of right now is use the current strict isolation nft rules to mark packets instead of dropping them, and then use an ip rule based on this mark to sent traffic to default gateway.

trozet replied:
@pliurh I think we still need to figure this one out

trozet replied:
yes, sorry, I had it backwards. @pliurh can you make sure we have a section in the enhancement covering this?

jcaamano commented:
> > > so like the no-overlay mode for default network also rely on the BGP ( frr-k8s) , in case frr-k8s is stop working or not stable. The traffic between different nodes will also be affected. but I hope my my concern is unnecessary :)
> > 
> > 
> > I guess frr-k8s becomes a point of failure equivalent to ovn-k. I am not sure if we would need to take additional considerations to ensure that frr-k8s being down does not disrupt existing networking confuration. @fedepaol WDYT?
> 
> If FRR-K8s becomes part of the control plane of east - west traffic, when it goes down the routes will be removed from the kernel and east - west with that node won't work. Same happens with the router, just I'd trust the router more than something I wrote :P
> 
> The only thing to alleviate I can think of, is the addition of graceful restart which frr-k8s supports (cc @karampok , please keep me honest here). Any "manual" preservation of the routes (ie ovnk monitors and installs the route if frrk8s goes down) is risky imo and can be worse if not handled correctly

I see. I would have thought there would be a configuration mode which you would tell FRR "hey, if anything goes wrong, just keep things as they were until things go right again"

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
````diff
@@ -0,0 +1,840 @@
+      nodeSelector: {}
+    ```
+
+4. The cluster admin enables no-overlay mode by running ovn-kubernetes with the
+   configuration flag `--default-network-encapsulation=none` set.  
````
No. If the flag is unset. OVN-K will operate in the geneve overlay mode by default.

pliurh replied:
Yes. The ovnkube-node pod shall exit with an error.

zhaozhanqi replied:
do we need to consider OCP cluster here. Like what the steps need to prepare for day1 ? how to set the value to enable the no-overlay?  the external router is must before deploying the no-overlay cluster?  or create an internal Frrconfiguration with all nodeip as the neighbors if not external router . i.e

```
spec:
  bgp:
    routers:
    - asn: 64512
      vrf: blue
      neighbors:
      - address: 192.168.111.20
        asn: 64512
        toReceive:
          allowed:
            mode: all
      - address: 192.168.111.21
        asn: 64512
        toReceive:
          allowed:
            mode: all
```
  cno trigerr the frr-k8s and ovn-k with RA and frrconfiguration?  otherwise I guess the container pods cannot be communicate with different nodes. like api and auth operator may failed

pliurh replied:
@zhaozhanqi that can be addressed in the d/s enhancment. frr-k8s pods can be deployed in hostnetwork.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
````diff
@@ -0,0 +1,840 @@
+    Note over Node B: Traffic is delivered to Pod2
+```
+
+To maintain SNAT for non-BGP-routed egress traffic, in the gateway router, the
+default SNAT rule will be kept. The route import controller will add a Logical
+Router Policy (LRP) to the gateway router for each BGP imported route to skip
+the SNAT rule.
+
+In the following example, `10.128.0.0/24` is the local nodeSubnet,
+`174.16.1.0/24` is the CIDR of a BGP route.
+
+```
+100 ip4.src == 10.128.0.0/24 && ip4.dst == 174.16.1.0/24         outport                rtoe-GR_ovn-worker
+```
````
After some experiments, it appears that LRP cannot skip SNAT. I also tried to set the `exempted_ext_ips` flag in SNAT rules. It did not function as outlined in the OVN documentation. Do you know of any ways to implement this function in SGW mode?

jcaamano replied:
Why not conditional SNAT as I mentioned before?

jcaamano replied:
> > There could be another route with a longer prefix match that is not BGP.
> 
> I cannot think of a good solution for this case. The ideal approach is to bypass SNAT based on the nexthop IP. But OVN cannot do that.

Let's see what others think.

We could have a BGP for no-overlay only mode, where we just don't SNAT internal E/W traffic.

I guess it would not be complicated to calculate the routing space to come up with the real destinations that are BGP. However not sure such an effort should be scoped in the no-overlay epic if there are other acceptable alternatives.

karampok commented:
> > > > so like the no-overlay mode for default network also rely on the BGP ( frr-k8s) , in case frr-k8s is stop working or not stable. The traffic between different nodes will also be affected. but I hope my my concern is unnecessary :)
> > > 
> > > 
> > > I guess frr-k8s becomes a point of failure equivalent to ovn-k. I am not sure if we would need to take additional considerations to ensure that frr-k8s being down does not disrupt existing networking confuration. @fedepaol WDYT?
> > 
> > 
> > If FRR-K8s becomes part of the control plane of east - west traffic, when it goes down the routes will be removed from the kernel and east - west with that node won't work. Same happens with the router, just I'd trust the router more than something I wrote :P
> > The only thing to alleviate I can think of, is the addition of graceful restart which frr-k8s supports (cc @karampok , please keep me honest here). Any "manual" preservation of the routes (ie ovnk monitors and installs the route if frrk8s goes down) is risky imo and can be worse if not handled correctly
> 

When in a BGP peering between peer A  and  B we could have
- Peer A advertising GR to B, so B will keep the routes  from A when BGP session goes down
- Peer A must start zebra daemon with an argument to keep all the routes (BGP routes mainly) until zebra restarts

In theory Graceful Restart can help but the are more risks than benefits. E.g. if the node reboots, the local frr process will trigger GR so peer will keep routes towards the node, those routes  will stay for a 2 minutes but then traffic will be blackholed because the dataplane went eventually down. (BFD are ignored during)

> I see. I would have thought there would be a configuration mode which you would tell FRR "hey, if anything goes wrong, just keep things as they were until things go right again"

We need a definition of what "goes wrong" mean, and what "things go right" mean.

What is hard it to actually say "As long as the dataplane works, keep things as they are", frr process can only know status about control plane, and the control-plane/data-plane have different status (e.g. frr process crashes, but data works).  In theory there is a BFD configured to work dataplane level  that could help (but I have not looked into)

jcaamano commented on docs/okeps/okep-5259-no-overlay.md:
````diff
@@ -0,0 +1,840 @@
+    Note over Node B: Traffic is delivered to Pod2
+```
+
+To maintain SNAT for non-BGP-routed egress traffic, the masquerade iptables rule
+will be kept. The route import controller will insert a iptables rule for each
+BGP imported route to skip the SNAT rule to the gateway router.
+
+In the following example, `10.128.0.0/24` is the local nodeSubnet,
+`174.16.1.0/24` is the CIDR of a BGP route.
+
+```
+-A POSTROUTING -s 10.128.0.0/24 -d 174.16.1.0/24 -j RETURN
+-A POSTROUTING -s 10.128.0.0/24 -j MASQUERADE
+```
````
I am not sure the route_import controller is the best place to do this. That controller is specific to OVN and already exceptional in that is IC only, both concerns which do not apply to LGW mode.

jcaamano replied:
I think a new controller could be better. The only shared responsibility would be to subscribe to route updates and I am not sure that's enough to justify a mix.

jcaamano commented:
> We need a definition of what "goes wrong" mean, and what "things go right" mean.

Right.

So let's say I loose connection with peer A. Until I recover the connection, let's keep the routes I learned from A.
Every other problem is a derivative. Let's say I crash and restart: until I connect to A, let's keep the routes I learned from A.

So I guess the though model is: once I configure the data plane in a certain way, I don't change that configuration unless I know for a fact I should.

That's a similar thought process as how OVS configuration works, I can tell it what to do it it loses connection to the controller: open-fail mode -> keep flows as they are, secure-fail mode -> clean up flows installed by controller

ricky-rav commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,889 @@
+    apiVersion: frrk8s.metallb.io/v1beta1
+    kind: FRRConfiguration
+    metadata:
+      name: cudn-1
```
nit: maybe let's call it `no-overlay-default` or something similar?

ricky-rav commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,889 @@
+##### Internal Route Reflectors
+
+While an FRR instance can act as a BGP route reflector, enabling an internal
+route reflector-based BGP network across the cluster. However, it is not
```
Is there a verb missing in the main sentence here?

ricky-rav commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,889 @@
+add the new `encapsulation` field, it will be available for both CUDN and UDN.
+So, to prevent a user from setting this field in a UDN CR accidentally, a new
+`status.conditions` type `NoOverlayEnabled` will be added to CUDN/UDN CRs. If a
+user tries to create a UDN CR with no-overlay mode enabled, cluster-manager
```
So did we dismiss the CEL in the end?

pliurh replied:
CEL is part of the API. We want to support no-overlay UDN in the future, once BGP supports UDN. With this `status.conditions` approach, we don't need to change the API(CEL) when we support UDN. @jcaamano @tssurya WDYT?

jcaamano replied:
For me CEL still looks like the better option

ricky-rav replied:
Yeah, so the CEL will be removed once we support no-overlay mode for UDN. Usually it's more confusing for a user to see their CR created and then realize that the cluster doesn't work as expected, rather than having the CR rejected from the start with a descriptive error message.

tssurya replied:
> CEL is part of the API. We want to support no-overlay UDN in the future, once BGP supports UDN. With this `status.conditions` approach, we don't need to change the API(CEL) when we support UDN. @jcaamano @tssurya WDYT?

yea having a CEL now to be removed later is fine

ricky-rav reviewed (commented):
Just a few minor comments.

tssurya commented on docs/okeps/okep-5259-no-overlay.md:
````diff
@@ -0,0 +1,882 @@
+            disableMP: true
+    ```
+
+3. The cluster admin shall create a RouteAdvertisements CR to advertise the
````
hmm am I confusing something? I don't think creating RA CR is a requirement for no-overlay mode right? I think perhaps you mean to update the enhancement around this ? like we discussed in BGP sync meeting I think we want to create things internally for the user rather than expect them to have RA created etc.. which has a totally different API meaning.
I assumed RA today means:

1. advertise pod networks to outside the cluster elements

and this meaning of the API mustn't change IMHO

1. advertise pod networks to other nodes in the cluster -> this is the semantic of the new field you are adding which should require no additional end-user API changes

tssurya commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,882 @@
+advertised to the same VRF, the cross-UDN traffic is still blocked. This is the
+current behavior when a network is advertised. We will not change it for no
+overlay mode.
+
```
In the implementation section please flesh out what are the changes coming into:

1. cluster-manager
2. ovnkube-controller
3. ovnkube-node

components in detail - so either you can call them out in the above pieces or write them separately

tssurya commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,882 @@
+
+#### BGP Topology
+
+The selected `FRRConfiguration` CR determines the deployment mode for the
+no-overlay network. There will be no new field added to the FRRConfiguration.
```
I think we are currently using RA API as the link to make changes - for me that's an extra user step.. shouldn't we do the configs ourself internally if the no-overlay-mode flag is set to none for that network?

tssurya reviewed (changes requested):
Great enhancement read! Thanks for the awesome writing.
My main points are (also covered as comments);

1. its good to flesh out the workflow for only no-overlay-mode w/o RA i.e w/ SNAT mode in workflow section
2. missing API types.go changes/future extensibility/CEL validation rules considerations - status feedback info to the users - what new condition types if any?
3. missing details on what are the changes in controllers to cluster-manager, ovnkube-controller, ovnkube-node component wise split - some parts are interleaved in sections but good to call them out
4. != matches are not that performant IIUC and in connecting UDNs enhancement we are going to move those cSNATs to the node as NFT rules to ensure h/w offload doesn't break for asymmetric traffic there
5. Scale/Perf section should be added to call out risks - example like if its a cluster with large number of nodes perhaps using the route reflectors makes more sense for performance? how will the controller changes impact setup/scale
6. Are we going to consider allowing for a scale/perf improvement around -> if a given UDN network only spans across 10 nodes in the cluster out of 500 then do peering only for those 10 nodes? but i think this is a larger community effort as was discussed in meetings and something in future

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,888 @@
+  visibility.
+* Leverage Existing Network Infrastructure: Integrates more seamlessly with
+  existing BGP-capable network devices, allowing for direct routing to pod IPs.  
+* Reduced Resource Consumption: Fewer CPU cycles spent on
```
The statement here is "Fewer CPU cycles spent on encapsulation/decapsulation". That is 100% true, as there is no enca/decap anymore. I am not saying the overall CPU consumption will be less. The overall CPU consumption can be affected by many factors (route numbers, packet size, packet rate...). So it's hard to say whether the CPU cycles we save from enca/decap are more than the ovnk controller changes consumed.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,882 @@
+
+## Goals
+
+* Support no-overlay mode for the default network.
```
No-overlay will only be available for BGP-advertised networks. More details can be found in the subsequent chapters.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,888 @@
+* Support EgressIP and EgressService features in networks operating in
+  no-overlay mode. Both features leverage OVN-Kubernetes's logical routing
+  policy to steer pod egress traffic. They also rely on overlay tunnels to
+  transport pod egress traffic to egress nodes. Implementing these features in
```
In the no-overlay mode, when the pod outbond traffic leaves its node, we have no control over how the next-hop routers forward the traffic, unless the nexthop is an ovn-k node. Therefore, we cannot ensure that the egress traffic can be forwarded to the egress IP node. That is the main difficulty of supporting EIP.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,882 @@
+  simultaneously. It is useful for scenarios such as overlay mode migration or
+  hybrid provider networks.
+
+## Non-Goals
```
It depends on whether route advertisements will be available for tenants or not.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,882 @@
+  to support no-overlay mode.
+
+* This enhancement does not aim to change the existing CUDN lifecycle
+  management. The user must ensure that the CUDN CRs are correctly managed
```
This means how users shall create/delete a CUDN. @ricky-rav

ricky-rav replied:
yeah, that's the idea.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,882 @@
+  cluster running in interconnect mode.
+
+* This enhancement does not aim to implement no-overlay mode for the layer-2 or
+  localnet type of networks. The BGP feature doesn't support the localnet type
```
It's about the CUDN configuration. Users shall not configure a CUDN in localnet with no-overlay enabled.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,882 @@
+  through a layer-2 segment using Geneve encapsulation. It's quite challenging
+  to implement a layer-2 network over a layer-3 infrastructure without an
+  overlay protocol.
+
```
We allow advertising CUDNs with overlapping pod subnets, as long as they are not advertised to the same VRF. This is the limitation of the BGP feature. If a user wants to turn on no-overlay mode for overlapping pod subnets, they shall configure a VRF-lite topology.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,882 @@
+### Story 3: Enable no-overlay mode without an external BGP router
+
+As a cluster admin, I want to avoid all encapsulation for traffic in the
+cluster, but I don't want to deploy an external BGP router for the cluster.
```
This is just a deployment option for users who need no-overlay, but don't want to advertise the pod network to the external.
There is no 'default way' for this feature. Users shall choose the deployment mode based on their infrastructure and requirements.

ricky-rav replied:
To integrate Peng's words:
```suggestion
cluster, but I don't want to advertise the pod network to the external and don't want to deploy an external BGP router for the cluster.
```

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,882 @@
+  who only want to have no-overlay mode but do not want to expose podIPs to the
+  external, this behavior is not ideal. Instead of disabling SNAT for all the
+  pod-to-external traffic, OVN-K shall adjust its SNAT behavior so that pod IPs
+  are exposed for traffic routed via BGP, but SNAT is still applied for
```
No-overlay mode can only be set when a network is advertised. However, a network is advertised doesn't mean the network is advertised to all external networks. For example, a node can have a default route to a non-BGP router, and advertise the pod network to another BGP router. In this case, the outbond traffic to default route shall remain SNATed whereas the traffic to BGP router shall not be SNATed.

ricky-rav replied:
Uhm how do you tell whether the default route goes via a BGP router or not?

pliurh replied:
If a route is learnt from a BGP router. I assume the nexthop is a BGP router.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,882 @@
+We need to update the ZoneInterconnectHandler for no-overlay mode. If a new
+network is configured in no-overlay mode, ZoneInterconnectHandler shall not
+create the transit switch for the network. For each remote zone, there is no
+need to add the static nodeSubnet routes into the ovn_cluster_router. The
```
IIRC, @jcaamano suggested removing the transit switch for a clear topology.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,882 @@
+
+The FRR instance on each node maintains a full mesh BGP peer relationship with
+all other nodes across the cluster. In this mode, users can enable no-overlay
+mode without relying on external BGP routes.
```
The workflow is the same as other deployments. The sole distinction lies in the BGP topology, as specified by the FRRConfiguration CR. I've given an example below.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,882 @@
+either `none` or `geneve`, with `geneve` being the default. Setting
+`--default-network-encapsulation=none` will configure the default network to
+operate in no-overlay mode. If the `--default-network-encapsulation=none` flag
+is set but the required RouteAdvertisements or FRRConfiguration CR is missing,
```
This is to define the behavior when ovnkube starts `--default-network-encapsulation=none` but without `--enable-route-advertisements`.

zhaozhanqi replied:
@pliurh 
> This is to define the behavior when ovnkube starts --default-network-encapsulation=none but without --enable-route-advertisement

is this mean if no-overlay is enabled for default network, but no need to --enable-route-advertisements ?  there will no routes in nodes host or GR router IIUC, is that expected?

zhaozhanqi replied:
or as long as no-overlay enabled,  there will the related BGP routes generated in nodes host and GR router, then no need to apply RA cr?

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
````diff
@@ -0,0 +1,882 @@
+      x-kubernetes-validations:
+      - message: Setting layer3.encapsulation is not supported for UserDefinedNetworks
+        rule: '!has(self.encapsulation)'
+```
````
I'll add a status condition for that.

pliurh replied:
Added

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,882 @@
+gateway_port        : []
+logical_ip          : "10.244.1.7"
+logical_port        : []
+match               : "ip4.dst!=$skip_snat_dst||ip4.dst==$snat_dst"
```
`$skip_snat_dst` contains the prefixes of the imported BGP routes. I knew there would be a performance challenge. But I cannot think of a better approach.

ricky-rav replied:
If this is really a performance bottleneck, @jcaamano was suggesting on slack that we could derive the subnets for which we want to SNAT and apply it only to those:
https://redhat-internal.slack.com/archives/C090C97A90B/p1751622131107829?thread_ts=1751450578.459089&cid=C090C97A90B

I'm repeating this here because I've found it by chance and discussions are already scattered in multiple places :)

pliurh replied:
That is the default behavior today for the overlay mode. So users don't need to configure NAT at the gateway router of the cluster if it's not a BGP router. It's also the default behavior of Calico BGP mode.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,882 @@
+The OVN-Kubernetes BGP feature currently disables SNAT only for traffic destined
+to external networks, not for inter-pod communication across nodes.
+
+In no-overlay mode, the goal is to fully leverage BGP and eliminate the SNAT
+step on the gateway router, even for intra-cluster traffic. Within the same
```
Pod network will always be advertised in the no-overlay mode.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,882 @@
+external_mac        : []
+external_port_range : ""
+gateway_port        : []
+logical_ip          : "10.244.1.7"
```
Yeah, it shall be from 10.128.0.0/24

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,882 @@
+Currently, BGP routes to other node subnets are not imported into the gateway
+router, which ensures node-to-pod traffic traverses the OVN logical topology.
+However, we need the routes to be imported into the gateway router in no-overlay
+mode. We need to update the route import controller to check whether a network
```
> if we fix the host-routing-table side via FRRConfig

I don't know what you mean by this.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,882 @@
+UDNs or between UDNs and the default network. So, even if multiple UDNs were
+advertised to the same VRF, the cross-UDN traffic is still blocked. This is the
+current behavior when a network is advertised. We will not change it for no
+overlay mode.
```
The UDN isloation logic will not be affected by no-overlay mode with or without https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5276

ricky-rav commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,882 @@
+  are exposed for traffic routed via BGP, but SNAT is still applied for
+  non-BGP-routed egress traffic.
+
+* For east-west traffic, intra-node traffic (pod-to-pod, pod-to-clusterIP, and
+  pod-to-host) remains unchanged, while cross-node traffic will follow the same
+  path as north-south traffic.
```
I think it's important to make this distinction extra clear in the document. 
So here we're saying that for clusters with no-overlay mode where the admin doesn't want to expose the pod subnets to the outside, node SNAT is still in place for egress traffic. Is this correct?

maiqueb commented:
/cc

trozet commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1054 @@
+(like a default internet gateway). So, instead of disabling SNAT for all pod
+egress traffic, we shall do the following:
+
+* For egress traffic forwarded via a BGP route, SNAT shall be disabled.
+* For egress traffic forwarded via a non-BGP route, the egress traffic shall be
+  SNATed to the node IP.
```
This seams really complicated to pull off. Is it really a hard requirement? Could we just expose a field in the RA that says "SNAT egress traffic" or something? If true, then we only add pod subnets to the address set and skip SNAT if they are in there (positive match), all other traffic is SNAT'ed.

trozet commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,882 @@
+
+### Deployment Consideration
+
+#### BGP Topology
```
+1 this is a great section now, with awesome diagrams. Thanks!

trozet commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1054 @@
+All the limitations of the BGP integration feature also apply to the
+no-overlay feature.
+
+As OVNI is no longer delivering pod-to-pod traffic end-to-end, it will
```
nit: s/OVNI/OVN

trozet commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,882 @@
+
+#### Multiple External Gateways (MEG)
+
+The same as the route advertisement feature.
```
I think she means the -ds argument to kind:
```
-ds  | --disable-snat-multiple-gws            Disable SNAT for multiple gws. DEFAULT: Disabled.
```

If -ds is not used, then by default there is one SNAT for the entire pod subnet on the GR. If -ds is used, we create per pod SNATs. -ds is defaulted to true in OCP and required to use MEG.

pliurh replied:
MEG cannot work with advertised networks, therefore MEG cannot work with networks operates in the no-overlay mode either.

ricky-rav reviewed (commented):
Looks very good. Thanks @pliurh for making all assumptions explicit throughout the document and describing the reasoning behind each decision.

maiqueb reviewed (changes requested):
tks for this proposal.

ricky-rav commented on docs/okeps/okep-5259-no-overlay.md:
````diff
@@ -0,0 +1,1054 @@
+the following CEL validation rule to the UserDefinedNetwork CRD.
+
+```yaml
+    layer3:
+      description: Layer3 is the Layer3 topology configuration.
+      x-kubernetes-validations:
+      - message: Setting layer3.encapsulationConfiguration is not supported for UserDefinedNetworks
+        rule: '!has(self.encapsulationConfiguration)'
````
the rule translates into: "this UDN is valid if it doesn't have `encapsulationConfiguration` set"

maiqueb replied:
Maybe I wasn't clear enough, sorry for that. I fully understand what this CEL expression does. What I do not understand is how this CEL expression can be used to achieve the goal you have set, which, IIUC, is to allow `encapsulationConfiguration` to only be available on C-UDNs.

Using your own words: since the `Layer3Config` is shared between CUDN and UDN, once you add the new `encapsulationConfiguration` field how can this CEL expression - which is mapped to `Layer3Config` - prevent adding an `encapsulationConfiguration` to a UDN, but allow it for a C-UDN ?

Surely I'm missing something.

pliurh replied:
This CEL is only added to UDN. So it means in a UDN CR is valid if it doesn't have encapsulationConfiguration set. It prevents user from setting `encapsulationConfiguration` for a UDN.

pliurh replied:
This CEL will only be added to the UDN CRD. So the CUDN will not be affected.

maiqueb replied:
Ah, OK. Does that the `Layer3Config` will no longer be shared between CUDN and UDN - each will have different attributes ? If so, IMHO, your text should be updated. 

If not, would you please explain how can you define CEL expressions on a shared type that will be added to only one of the CRDs ?

maiqueb replied:
Right, thank you for literally pointing it out.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
````diff
@@ -0,0 +1,1054 @@
+gateway router according to source-based route:
+
+```
+IPv4 Routes
+Route Table <main>:
+                10.96.0.1                22.100.1.2 dst-ip
+               10.96.0.10                22.100.1.2 dst-ip
+               100.65.0.3                100.65.0.3 dst-ip
+            22.100.1.0/24                100.65.0.3 src-ip
+            22.100.0.0/16                100.65.0.3 src-ip
+```
````
This is the routing table of ovn_cluster_router which doesn't contains any BGP routes. I put it here to show that thte pod-to-remote-pod traffic will not be routed to the transit swith in the no-overlay mode.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1054 @@
+
+For CUDNs, no-overlay mode shall be configured via the CUDN CRs.
+
+### API Details
```
Here's the assumption: 
1. If the egress traffic is routed via an external BGP route, it implies that the pod subnet is learnt by the external network. So we don't need SNAT to the node IP.
2. If the egress traffic is routed via a non-BGP route, it implies that the pod subnet is unknow to the external network. So we shall SNAT it to the node IP.

maiqueb replied:
Right - I do understand that. 

What I wonder is, if the assumption will match the expectations of the entire user base. Hence, I would suggest to do nothing, and offload this to a separate enhancement.

Point being, I'd rather rely on user feedback than on assumptions - which are often disastrous. 

Just a thought.

pliurh replied:
Added a new field to RA to make it configurable.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1054 @@
+  pods have a per-node logical switch and subnet. Routing is used for pod to pod
+  communication across nodes. This is the network type used by the cluster
+  default network today.
+* **Localnet Type Network** - An OVN-Kubernetes topology rendered into OVN where
```
I copied this line from https://github.com/pliurh/ovn-kubernetes/blob/882f4451fc0e30973b03a7f1009b80d7b5e854b6/docs/okeps/okep-5193-user-defined-networks.md#L31-L32.
And according to https://github.com/ovn-kubernetes/ovn-kubernetes/blob/d56ed1edf7a9705cf8e579b40d50cc2bfdec560b/go-controller/pkg/ovn/base_secondary_layer2_network_controller.go#L181, there is no transit switch for a Localnet network.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1054 @@
+            "mtu": 1500,
+            "netAttachDefName": "default/l3-network",
+            "role": "primary",
+            "encapsulation": "none"
```
Users shall not set the encapsulation feild in a NAD, it shall be generated from a CUDN CR by OVN-K.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1054 @@
+
+#### UDN Traffic Isolation
+
+Currently, for cross-UDN pod-to-pod traffic, OVN-Kubernetes supports 2 modes:
```
https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5276

maiqueb commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1054 @@
+## Goals
+
+* Support no-overlay mode for the cluster default network(CDN).
+* Support no-overlay mode for Primary layer-3 ClusterUserDefinedNetworks
```
OK; then please add that to the enhancement.

ricky-rav commented on docs/okeps/okep-5259-no-overlay.md:
````diff
@@ -0,0 +1,1079 @@
+  // +optional
+  // +kubebuilder:default=false
+  NatOutbound bool `json:"natOutbound,omitempty"`
+```
````
We discussed this briefly on Monday during the BGP call and I don't have a solid opinion about this yet. But here's a few thoughts.

First of all, I think we need a wide consensus on the fact that we need `NatOutbound`. I do prefer this approach to the previous one, where the SNAT behaviour depended on the route a packet would hit on egress: the behaviour is clear from the start and can be decided by the cluster admin.

As for where this parameter should be, I think it makes sense to have all parameters for the same feature grouped together, wherever possible. Does it make sense for `NatOutbound` to be configured in the RA and the no-overlay mode to be enabled 1) as an input parameter for the default network 2) in the CUDN spec for whatever CUDN we want to affect? 
Would it make more sense to have it on a per-network basis instead, along with the field(s) that enable no-overlay? 
If the naming `encapsulation` that we chose for the spec makes it difficult to add `NatOutbound` underneath, maybe we can rename it to something else, like `OverlayMode` and model the values accordingly. 

It's just a thought. If it makes more sense to have it under RA, I'm fine with that too!

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1104 @@
+    // +kubebuilder:validation:Enum=None;Geneve
+    // +kubebuilder:default=Geneve
+    // +optional
+    Overlay OverlayProtocol `json:"overlay,omitempty"`
```
The proposed design follows the same pattern as the `topology` config in the UDN CRDs. And it aligned with the [Kube API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#automatic-resource-allocation-and-deallocation).

To support EVPN, it would be like this:
```
    topology: Layer3
    layer3:
      role: Primary
    overlay: EVPN              # Geneve | None | EVPN
    evpnOptions:
        evpnRef:
            name: blue-evpn     # name of the EVPN CR that defines macVRF/ipVRF
```

trozet replied:
ok makes sense

trozet commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1104 @@
+  who only want to have no-overlay mode but do not want to expose pod IPs
+  externally, this behavior is not ideal. Instead of disabling SNAT for all
+  pod-to-external traffic, we shall let users to choose whether SNAT shall be
+  skipped for the pod egress traffic by adding a new `natOutbound` field to
```
So this is a use case where you want to use BGP just for east/west traffic, and have no external BGP peer for N/S. Sorry I missed that in your Story 3. If we want to include it in the enhancement then that is OK, but I still think the API needs some work. We can discuss that in a separate comment though.

trozet commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1104 @@
+    * `Disable`: Does not SNAT pod-to-external traffic. This requires pod IPs to
+      be routable on the external network.
+    * Pod-to-remote-pod traffic is never SNATed in no-overlay mode.
+  * `routeAdvertisements`: Specifies the name of the RouteAdvertisements CR used
```
>No, the RA here is only for advertising podNetwork for enabling no-overlay. It is not for configuration but more for cross-check between CUDN and RA.

I mean't a use case for enabling NAT without no overlay mode is when you only want to advertise EgressIPs and not the pod network with a BGP RA. Maybe that is already implied today with the current RA behavior. Need to check.

jcaamano replied:
> I meant a use case for enabling NAT without no overlay mode is when you only want to advertise EgressIPs and not the pod network with a BGP RA. Maybe that is already implied today with the current RA behavior. Need to check.

Yes, advertising just EIPs will keep the NAT in place.

For every option that we add to the RA we need to think how we are going to aggregate it if multiple RAs select the same network providing each different values of this option. NAT is something that we either do for a network or we don't do for a network. So on that scenario we need to either report an error or make an assumption: for example, as long as one of the RA request the NAT, then it stays regardless of what the other RAs are requesting. This requires specific handling on our side and it stays confusing for a user. 
Whereas a NAT option on the network itself seems more straight forward to understand I think.

jcaamano replied:
I expressed the idea of specifying an RA in the CUDN [here](https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5289#discussion_r2345388797):

```
Both the no overlay and disable snat options assume there is something else configured that would make that work, like for example an RA. We should be able to do that, just assume there is something and not be concerned about it. However I have been thinking it could be beneficial for the user to specify what those RAs are. That could be like a three-way acknowledgement between the RA, the network and the user that the configuration that needs to be in place, is indeed in place. Additionally, we could use the references to those RAs to perform additional validation and report network scoped problems in different ways: failing to create the network, network status conditions, network events...
```

So it is basically used for validation. If we don't like the idea, we can drop it and just trust the user is doing what it needs to be doing.

jcaamano replied:
That's a question that wouldn't need to be answered if we had an API that promoted acknowledgement and validation :tongue: 

In any case, I understand how it can be confusing and I am ready to drop my suggestion and move on.

pliurh replied:
> What's the consequent state we will be in if the RA doesn't exist selecting the network but user created a no-overlay network?

We shall still allow the network to be created. The north-south traffic works, whereas the east-west routing doesn't. It allows users to create RA after CUDN or switch an RA for a network.

Currently, the advertisement controller sets the annotation `k8s.ovn.org/route-advertisements` to NADs. It shall also set the annotation to CUDNs. Then the CUDN controller can use it as reference to validate the RA configuration and check the RA status for no-overlay mode.

tssurya replied:
> Currently, the advertisement controller sets the annotation `k8s.ovn.org/route-advertisements` to NADs. It shall also set the annotation to CUDNs. Then the CUDN controller can use it as reference to validate the RA configuration and check the RA status for no-overlay mode.

ah! good point - @pliurh is this mentioned in the implementation details? perhaps I missed it.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1104 @@
+* When OVN-Kubernetes supports BGP for UserDefinedNetwork CRs, extend no-overlay
+  mode support for Primary layer-3 UDNs.
+
+* Support toggling no-overlay mode on/off for an existing network. To support
+  this functionality, there are two things we need to consider.
+  First, in no-overlay mode, as there is no Geneve overhead, a network can have
+  the same MTU size as the provider network. However, overlay networks require a
+  smaller MTU (default 1400). We cannot modify a pod interface MTU without
+  restarting the pod. So OVN-Kubernetes must work with other external
+  controllers to accommodate such a migration.
+  Secondly, to facilitate a smooth, node-by-node migration, it's crucial to
+  allow networks to operate in a hybrid no-overlay and overlay mode.  
```
> If users would be willing to accept temporary disruption, would this alleviate the concern?

Yes. But we'd better add a status indicator to the CUDN API to let users know when the transition is complete.

I think the options to bypass MTU migration are valid. However, until we get more precise customer requirements, I believe it would be best to leave it as a future goal.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,888 @@
+* Improved Performance: Eliminates encapsulation/decapsulation overhead,
+  potentially leading to lower latency and higher throughput for inter-pod
+  communication.
+* Simplified Troubleshooting: Traffic paths are more transparent as they align
```
Well, that's true. I'll remove this paragraph.

tssurya commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1105 @@
+ClusterUserDefinedNetwork (CUDN) CRD to control network encapsulation:
+
+* **`overlay`**: Specifies the encapsulation method.
+  * Supported values: `Geneve` (default) or `None` (no-overlay).
```
@pliurh : thinking out loud.. on calling it `no-overlay` -> today we assume its no-overlay with BGP since scope of RA is just for BGP - although nothing prevents us from adding OSPF tomorrow and the name `RouteAvertisement` is generic enough that I don't know if we might support more protocols. Or maybe it would be new API `LinkStateAdvertisement`
In that case, if we keep no-overlay as the name, then inside noOverlayOptions I hope we can have enough granularity for each type of no-overlay? OR do we want to be more specific around this?

ricky-rav commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1104 @@
+      - message: "noOverlayOptions is only supported for no-overlay networks"
+        rule: 'self.overlay == "None" || !has(self.noOverlayOptions)'
+      - message: "noOverlayOptions.routeAdvertisements is required when overlay is 'None'"
+        rule: 'self.overlay != "None" || (has(self.noOverlayOptions) && has(self.noOverlayOptions.routeAdvertisements) && self.noOverlayOptions.routeAdvertisements != "")'
```
Uhm... if we make this `RouteAdvertisements` field required, what value should it have if we want to support the simplest use case we discussed, where the cluster admin only wants to enable no-overlay for all eligible networks between nodes and doesn't want to export pod subnets to the outside? In this use case we're in charge of creating the base FRRConfig and the RA.

ricky-rav commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1104 @@
+      networkSelectors:
+      - networkSelectionType: DefaultNetwork
+      # The empty nodeSelector selects all nodes. We don't support a network in a overlay and no-overlay hybrid mode.
+      nodeSelector: {}
```
Do we report an error somewhere if the nodeSelector isn't empty?

pliurh replied:
ovnkube-cluster-manager shall print error to its log.

ricky-rav commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1104 @@
+
+### Workflow
+
+#### Enable No-overlay Mode for the Default Network
```
So this is the complete workflow for external reflectors. Did we drop the idea of internal route reflectors that we evoked two weeks ago in order to avoid a full mesh? Or are we leaving this to the downstream usage of the feature? We also talked about having the AS number configurable, in a similar way that calico does.

jcaamano commented on docs/okeps/okep-5259-no-overlay.md:
````diff
@@ -0,0 +1,1424 @@
+      - networkSelectionType: DefaultNetwork
+      nodeSelector: {}
+    ```
+
````
Yeah, but that's an important change because then we would have to generate FRRConfigurations per node with a node selector.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1424 @@
+
+Here is a configuration example:
+
+1. The frr-k8s pods shall be deployed on each node.
```
It truly doesn't matter. Let's leave it as it is.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1054 @@
+
+N/A
+
+## References
```
For now, EVPN integration is not in the scope of this enhancement.

jcaamano commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1491 @@
+* Address the advertised UDN isolation loose mode issue when no-overlay mode is
+  enabled.
+
+## Non-Goals
```
Not sure why we removed the non-goal of (not) supporting (tenant) UDNs

pliurh replied:
It contradicts with future goal `When OVN-Kubernetes supports BGP for UserDefinedNetwork CRs, extend no-overlay mode support for Primary layer-3 UDNs.`

jcaamano commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1491 @@
+  controllers to accommodate such a migration.
+  Secondly, to facilitate a smooth, node-by-node migration, it's crucial to
+  allow networks to operate in a hybrid no-overlay and overlay mode.
+* Address the advertised UDN isolation loose mode issue when no-overlay mode is
```
I think @tssurya expressed interest in scoping this in

jcaamano commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1491 @@
+### Configuration File and Flags
+
+For the default network transport, the configuration file will have a new section
+`[default-network]`, with the following new flags:
```
There is already a section called `[default]` which contains, among other things, the MTU for the default network. Isn't that existing section appropriate to hold these parameters?

jcaamano commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1491 @@
+
+* The `no-overlay-outbound-snat` flag is a string flag to configure the SNAT
+  behavior for outbound traffic from pods on the default network. Supported
+  values are `enable` or `disable`. It is only used when `transport=no-overlay`.
```
Required? Or what is the default?

jcaamano commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1491 @@
+# per-pod SNAT rule in the Gateway Router (exclude pod-network destinations)
+_uuid               : cd9264de-f2c6-4528-8f53-c2ccb5b23c8d
+allowed_ext_ips     : []
+exempted_ext_ips    : ["a15397791517117706455"] # Exclude traffic to the pod Network
```
I would call out we are using this to avoid the negative match that would need to happen if we used `match` instead. That's the reason right?

jcaamano commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1491 @@
+the east-west transport protocol of the network. If a network operates in
+no‑overlay mode, this key is set to `noOverlay`. This information can then be
+passed into the NetInfo object and utilized by both the RouteAdvertisement
+controller and the route import controller of ovnkube-controller.
```
There is no RouteAdvertisement controller in ovnkube-controller, I think it is ok if you just mention the route import controller here

jcaamano commented on docs/okeps/okep-5259-no-overlay.md:
````diff
@@ -0,0 +1,1491 @@
+    ```
+
+3. The default network controller in ovnkube-cluster-manager will create a base
+   FRRConfiguration CR named `default` in the `frr-k8s-system` namespace, and a
````
No, the FRRConfiguration is created by whatever controller is in charge of creating the internal BGP fabric. ItWhy would it be the default network controller in ovnkube-cluster-manager?
Also I would use more specific names than just `default` as they can collide with other instances. For example the RA controller is using `ovnk-generated-blabla` to avoid collisions

jcaamano commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1491 @@
+  nodes with a well-known label `k8s.ovn.org/internal-bgp-role=route-reflector`
+  as the route reflector, and generate the FRRConfigurations for other nodes as
+  clients. This topology is more scalable for large clusters. We will work in
+  the K8S-FRR project to support the route reflector feature.
```
> `route-reflector` is in the scope of the enhancement. However, it will be supported once we change K8S-FRR. So `full-mesh` will be the first topology we implement.

If it is in scope, then we need to describe what is the plan to implement it FRR-K8s? I don't see the point otherwise. It is difficult to understand what's your idea of how things should behave if a user sets `route-reflector`? It wont work silently? Do you plan to return an error while it is not implemented? I don't understand why we are going down the more complicated path here instead of scoping it out.

jcaamano commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1459 @@
+field instead of adding a negative match to the `match` field, as using a
+negative match in an OVN SNAT rule is not efficient. Setting the
+`exempted_ext_ips` field results in one OpenFlow rule being created for each pod
+SNAT entry in the OVS bridge.
```
nit:

The phrase starts with `If `outboundSNAT` is enabled for a network` and that's important. If the network has geneve transport or `outboundSNAT` disabled, we shouldn't be doing any of this to avoid the risk of impacting scale. I am not sure if this could be conveyed more clearly. Same for local gateway mode where this is even less explicitly stated.

jcaamano commented on docs/okeps/okep-5259-no-overlay.md:
````diff
@@ -0,0 +1,1459 @@
+    ```
+
+4. The UDN controller in ovnkube-cluster-manager will create a
+   RouteAdvertisements CR named `blue` that advertises the blue network.
````
Again, maybe important to avoid generating common RouteAdvertisements CR names to avoid the potential for conflict.

maiqueb commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1459 @@
+In a large cluster, a full-mesh BGP setup leads to more CPU and memory
+consumption on the nodes. Instead of every node peering with every other node,
+nodes peer only with external BGP route reflectors. This significantly reduces
+the number of BGP sessions each node needs to maintain, improving scalability.
```
what does this have to do with no overlay ? I get why this is useful, and why ovn-k should want to implement it.
I just don't see what's the relationship of that with no overlay mode. 

@pliurh

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1491 @@
+  no `RouteAdvertisements` CR is configured for advertising the default network,
+  ovnkube-cluster-manager will emit an event and log an error message.
+
+* The `no-overlay-outbound-snat` flag is a string flag to configure the SNAT
```
I'll create the `[no-overlay]` section for these flags.

trozet commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1494 @@
+    OVN-Kubernetes will generate the BGP configurations (RouteAdvertisements and
+    FRRConfiguration) to advertise the pod subnets, and set up BGP sessions
+    between nodes. In the managed routing mode, pod subnets will be advertised
+    to the default VRF of each node for both the default network and CUDNs.
```
so this "managed" config flag enables managed for all the CUDNs as well? so it is not configurable in the CUDN?

pliurh replied:
This flag is for the default network only. CUDN gets its own flag in the CRD. This is a general statement for the managed mode. I should move this line somewhere else.

trozet commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1494 @@
+    external network. Therefore, the managed routing mode is not suitable for
+    clusters that have nodes in different subnets, as pod subnets are not
+    routable between nodes in different subnets. So normally, this option shall
+    be used with `outbound-snat=enable`.
```
should this be enforced that you outbound-snat must be enable when using managed mode in the CLI and in the CUDN?

trozet commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1494 @@
+  users. Supported values are `managed` or `unmanaged`. It is required when
+  `transport=no-overlay`.
+
+  * `managed`: OVN-Kubernetes manages the routing for the no-overlay network.
```
when managed mode is used, if on day 2 I decide I want to connect my workers to external BGP peers, can I do that with managed mode still enabled? Or do I need to switch to unmanaged? What happens if I switch? Does it delete all the frr-k8s configs or does it just stop managing them and leave them as is.

Can we update this okep with this information please?

trozet commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1494 @@
+   FRRConfiguration CR in the `frr-k8s-system` namespace, and a
+   RouteAdvertisements CR that advertises the default network. The generated
+   FRRConfiguration will configure BGP speakers on each node to advertise the
+   pod subnets and set up a full-mesh BGP topology where each node peers with
```
nit: full-mesh iBGP topology

just to be explicit

trozet commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1494 @@
+
+#### Multiple External Gateways (MEG)
+
+MEG cannot work with advertised networks, therefore MEG cannot work with
```
hmm I forgot I guess that we excluded MEG support with CDN + BGP. I guess you can ignore my earlier comments about MEG and we can handle feature support in a later OKEP.

trozet commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1494 @@
+    external network. Therefore, the managed routing mode is not suitable for
+    clusters that have nodes in different subnets, as pod subnets are not
+    routable between nodes in different subnets. So normally, this option shall
+    be used with `outbound-snat=enable`.
```
MEG not supported

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1492 @@
+
+The following diagram shows the OVN topology of a cluster with three networks:
+
+* a CUDN (in green) that uses Geneve encapsulation (default behavior)
```
Yes, that's a limitation of route advertisement, not the no-overlay mode. But you can use VRF-lite to isolate UDNs with VLANs.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
````diff
@@ -0,0 +1,1494 @@
+               100.64.0.4                100.64.0.4 dst-ip
+            10.128.0.0/24                100.64.0.4 src-ip
+            10.128.0.0/16                100.64.0.4 src-ip
+```
````
This is an existing route. It can work with no-overlay mode without a code change.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1494 @@
+        transport: "NoOverlay"
+        noOverlayOptions:
+          outboundSNAT: "Disable"
+          routing: "Unmanaged"
```
Yes

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1494 @@
+
+#### Egress IP
+
+Not supported.
```
Yeah, it cannot work if the nodes are not using each other as nexthop. So even if we implement EIP with no-overlay mode, there is still a limitation.

pliurh commented:
I updated the PR according to @trozet 's comments, and I also updated the support values of `outboundSNAT` from `enable`/`disable` to `enabled`/`disabled`

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1494 @@
+mode will use, the configuration file will have a new section `[bgp.managed]`, with the
+following new flags:
+
+* The `as-number` flag is used when no-overlay networks are configured with
```
The managed mode is a per-network setting for the no-overlay mode. The user may also want to advertise another network to external routers via the default VRF.

Another use case: the user can configure the network in managed mode (handling east/west traffic by the internal BGP fabric), and additionally create manual route advertisements for the same network to handle north/south traffic via external routers.

And yes, all such configurations can be archived using the unmanaged mode alone. However, I don't think we should prevent users from doing that in managed mode.

tssurya commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,888 @@
+  Secondly, to facilitate a smooth, node-by-node migration, it's crucial to
+  allow networks to operate in a hybrid no-overlay and overlay mode.  
+
+* Support EgressIP and EgressService features in networks operating in
```
@pliurh I think we dicussed this and agreed its too hard cause change needs to happen on the EgressIP and EgressService CRs which is complicated at this stage and out of scope and warrants its own epic? Was that reflected somewhere on the OKEP?

jcaamano reviewed (dismissed):
Some nits, LGTM overall.

Great effort @pliurh

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
````diff
@@ -0,0 +1,1494 @@
+    outbound-snat = disable
+    ```
+
+5. The default network controller in ovnkube-controller will create the OVN
````
I think it over. We cannot skip allocating `k8s.ovn.org/node-transit-switch-port-ifaddr` because it's shared by all networks. In a cluster, the default network can be in no-overlay mode, while a CUDN can be in overlay mode.

pliurh commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1523 @@
+For CUDNs, no-overlay mode is configured via their respective CRs. Due to a BGP
+limitation, CUDNs with overlapping pod subnets cannot be advertised to the same
+VRF. To enable no-overlay mode for these networks, a VRF-lite configuration is
+required.
```
I'll add this to the test plan.

jcaamano commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1523 @@
+### Testing Details
+
+* E2E Testing Details
+  The E2E tests shall cover the following combinations:
```
> CDN geneve <-> no-overlay transition is not supported.

hmm wouldn't this be the first thing we don't allow to reconfigure on restart?

Not sure if that's reasonable but @ricky-rav something to account for in your downstream enhancement,

tssurya commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,882 @@
+  Currently, BGP only supports advertising entire pod networks. In the future,
+  we may add support for selective node advertisement, which will enable hybrid
+  modes. This will allow networks to operate in overlay and no-overlay modes
+  simultaneously. It is useful for scenarios such as overlay mode migration or
```
> > do we also have a use case for connecting half nodes using tunnels and others in underlay using bgp as a permanent mode?
> 
> It may be useful once we have EVPN.

not sure I see how its useful BUT I see this part was removed and we only mention migration, so resolving this.

tssurya commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1554 @@
+  no-overlay mode but do not want to expose pod IPs externally, this behavior is
+  not ideal. To address this, users will be able to enable SNAT for
+  pod-to-external traffic. This will be configured via a new `outboundSNAT`
+  field in the `ClusterUserDefinedNetwork` CRD for CUDNs, and the
```
Just a note for my future self: That CUDN in itself when created is always going to SNAT pod to external to nodeIP.
What introduces the no-SNAT concept is the RA CRD, however the field to override that is being added via CUDN CRD.
Reasoning for this is because we want the granularity for this per network I guess. The network overlay itself definitely belongs to the network API, just the outboundSNAT is slightly misplaced but I am ok with the current proposal.

tssurya commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1554 @@
+To [maintain isolation between
+UDNs](https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5140), node BGP
+speakers do not accept routes to pod subnets by default. However, to route
+east-west traffic in no-overlay mode, these routes must be imported to the host
```
super nit: I'd clarify that we would thus import remote node pod routes *if no-overlay is enabled for a given network* and otherwise behaviour is unchanged.

tssurya commented on docs/okeps/okep-5259-no-overlay.md:
```diff
@@ -0,0 +1,1424 @@
+loose and strict. In the loose mode, cross-UDN traffic is allowed if an external
+router can route the traffic correctly. In strict mode, cross-UDN traffic is
+blocked by ACL rule defined in the node switch. In no-overlay mode, these two
+isolation modes shall still operate in the same way.
```
I think we agreed to make this a future goal thing right? So I can resolve this I think? and then https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5289/files#r2577174668
comment is same as this - there was some workaround @trozet had mentioned users can do at router config level for this.

pliurh commented:
> examples for UDN SNATs don't show the match...

The example is for `outboundSNAT=true` case, so we don't need the `match` field. Since all pod egress traffic will be SNATed except the pod-to-pod traffic.
